src/SparkFun_u-blox_SARA-R5_Arduino_Library.cpp

```diff
@@ -4693,6 +4693,76 @@ SARA_R5_error_t SARA_R5::ftpGetFile(const String& filename)
   free(command);
   return err;
 }
+SARA_R5_error_t SARA_R5::ftpChangeWorkingDirectory(const String& dirName)
+{
+  char *command;
+
+  command = sara_r5_calloc_char(strlen(SARA_R5_FTP_COMMAND) + dirName.length() + 16);
+  if (command==NULL)
+  {
+    return SARA_R5_ERROR_OUT_OF_MEMORY;
+  }
+
+  sprintf(command, "%s=%d,\"%s\"", SARA_R5_FTP_COMMAND, SARA_R5_FTP_COMMAND_CHANGE_DIR, dirName.c_str());
+
+  SARA_R5_error_t err = sendCommandWithResponse(command, SARA_R5_RESPONSE_OK_OR_ERROR, NULL,
+                                                SARA_R5_STANDARD_RESPONSE_TIMEOUT);
+  free(command);
+  return err;
+}
+
+SARA_R5_error_t SARA_R5::ftpCreateDirectory(const String &dirName) {
+
+  char * command;
+
+  command = sara_r5_calloc_char(strlen(SARA_R5_FTP_COMMAND) + dirName.length() + 16);
+  if (command==NULL)
+  {
+    return SARA_R5_ERROR_OUT_OF_MEMORY;
+  }
+
+  sprintf(command, "%s=%d,\"%s\"", SARA_R5_FTP_COMMAND, SARA_R5_FTP_COMMAND_MKDIR, dirName.c_str());
+
+  SARA_R5_error_t err = sendCommandWithResponse(command, SARA_R5_RESPONSE_OK_OR_ERROR, NULL,
+                                                SARA_R5_STANDARD_RESPONSE_TIMEOUT);
+  free(command);
+  return err;
+}
+
+SARA_R5_error_t SARA_R5::ftpList(const String &dirName,  char * response) {
```

Comment:
Consider renaming this to ```ftpListFiles```

Reply:
Humm. Thinking about this more...
The response could be quite large, depending on how many files are in the system. You could easily overrun response if you're not careful...
(I know I cheated with this overload of [getFileContents](https://github.com/sparkfun/SparkFun_u-blox_SARA-R5_Arduino_Library/blob/d8adcb3be61f3c658caa25b6896dc537da6b568e/src/SparkFun_u-blox_SARA-R5_Arduino_Library.h#L1011) because there you can know in advance how much data is coming.)
So, here, it would be best to use a String - which can grow as required - or pass in the size of response so you can stop writing before you overrun. String gets my vote - same as this overload of [getFileContents](https://github.com/sparkfun/SparkFun_u-blox_SARA-R5_Arduino_Library/blob/d8adcb3be61f3c658caa25b6896dc537da6b568e/src/SparkFun_u-blox_SARA-R5_Arduino_Library.h#L1010).

Reply:
Where possible I avoided the use of String in my changes to avoid heap problems, even though we're using an ESP32 so heap problems go away on deep sleep :)

I had originally changed a lot of existing methods to use char buffers too, but decided that was too much of an ask for a pull request so reverted them.

I'd be wary of using so many Strings on our Cortex M0+ based nodes which don't reboot after waking up.

You haven't found any problems with long running sketches?

Reply:
Hi David (@dajtxx ),
We haven't noticed any problems but: we also tend to run this on ESP32; and we generally don't put the processor to sleep. In this case you're welcome to twist my arm and go with a char buffer, with size.
Cheers,
Paul